core/amber/src/main/scala/edu/uci/ics/texera/web/service/ExecutionsMetadataPersistService.scala

```diff
@@ -57,16 +59,38 @@ object ExecutionsMetadataPersistService extends LazyLogging {
   ): Long = {
     // first retrieve the latest version of this workflow
     val uint = UInteger.valueOf(wid)
-    val vid = WorkflowVersionResource.getLatestVersion(uint)
+    val vid = getLatestVersion(uint)
+    val sid = getLatestSnapshot()
+    // get latest execution
+    val latestExecution = getLatestExecution(uint)
     val newExecution = new WorkflowExecutions()
     newExecution.setWid(uint)
     newExecution.setVid(vid)
+    // check if it's the first execution or different snapshot from previous
+    if (latestExecution.isEmpty || differentSnapshot(vid, latestExecution, uint)) {
+      newExecution.setSid(sid)
+    } else {
+      // delete new create snapshot
+      deleteSnapshot(sid)
+      // set current execution's snapshot to previous
+      newExecution.setSid(getLatestSnapshot())
+    }
+
     newExecution.setUid(uid.getOrElse(null))
     newExecution.setStartingTime(new Timestamp(System.currentTimeMillis()))
     workflowExecutionsDao.insert(newExecution)
     newExecution.getEid.longValue()
   }
 
+  // This function checks whether the snapshot is different from previous one
+  def differentSnapshot(
+      vid: UInteger,
+      latestExecution: Option[WorkflowExecutionEntry],
+      wid: UInteger
+  ): Boolean = {
+    !isSnapshotInRangeUnimportant(latestExecution.get.vId, vid, wid)
```

Comment:
no need for a function just for a single line, just call this line directly there

Reply:
refactored